Sources/NIOTestUtils/EventCounterHandler.swift

```diff
@@ -0,0 +1,363 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOConcurrencyHelpers
+
+/// `EventCounterHandler` is a `ChannelHandler` that counts and forwards all the events that it sees coming through
+/// the `ChannelPipeline`.
+///
+/// Adding `EventCounterHandler` to any point of your `ChannelPipeline` should not change the program's behaviour.
+/// `EventCounterHandler` is mostly useful in unit tests to validate other `ChannelHandler`'s behaviour.
+///
+/// - note: Contrary to most `ChannelHandler`s, all of `EventCounterHandler`'s API is thread-safe meaning that you can
+///         query the events received from any thread.
+public class EventCounterHandler: ChannelDuplexHandler {
+    public typealias InboundIn = Any
+    public typealias InboundOut = Any
+    public typealias OutboundIn = Any
+    public typealias OutboundOut = Any
+
+    public struct ValidityError: Error {
+        public var reason: String
+
+        init(_ reason: String) {
+            self.reason = reason
+        }
+    }
+
+    private let _channelRegisteredCalls = Atomic<Int>(value: 0)
```

Comment:
Jees, I really want some property delegates now to ease all of this boilerplate!